G2-http3-protocol.md

```diff
@@ -0,0 +1,137 @@
+H3_REQUEST_REJECTED(0x010b)|REFUSED_STREAM|UNAVAILABLE - Indicates that no processing occurred and the request can be retried, possibly elsewhere.
+H3_REQUEST_CANCELLED(0x010c)|CANCEL(8)|Mapped to call cancellation when sent by a client.Mapped to CANCELLED when sent by a server. Note that servers should only use this mechanism when they need to cancel a call but the payload byte sequence is incomplete.
+H3_REQUEST_INCOMPLETE(0x010d)|n/a|INTERNAL
+H3_MESSAGE_ERROR(0x010e)|n/a|INTERNAL
+H3_CONNECT_ERROR(0x010f)|CONNECT_ERROR|INTERNAL
+H3_VERSION_FALLBACK(0x0110)|n/a|INTERNAL
+n/a|FLOW_CONTROL_ERROR(3)|INTERNAL
+n/a|STREAM_CLOSED|No mapping as there is no open stream to propagate to. Implementations should log.
+n/a|COMPRESSION_ERROR|INTERNAL
+n/a|INADEQUATE_SECURITY| PERMISSION_DENIED … with additional detail indicating that permission was denied as protocol is not secure enough for call.
+
+#### Connection management
+
+`GOAWAY` and `PING` frames exist in HTTP/3 and serve the same purpose as in HTTP/2.
+One notable difference is the `GOAWAY` frame in HTTP/2 reports the last
+successfully processed stream ID. In HTTP/3 the `GOAWAY` frame ID value must be greater
+that the last successfully processed stream ID.
+
+### Exceeding deadlines
+
+When an RPC has exceeded its deadline, the server will reset the stream. In HTTP/2, a stream
+is reset using the `RST_STREAM` frame. The `RST_STREAM` frame doesn't exist in HTTP/3. 
+Instead, this action is performed using a QUIC frame, called `RESET_STREAM`.
+
+|                           | Frame        | Error code                   | Layer    
+----------------------------|--------------|------------------------------|----------
+Deadline exceeded HTTP/2    | RST_STREAM   | CANCEL(8)                    | HTTP     
+Deadline exceeded HTTP/3    | RESET_STREAM | H3_REQUEST_CANCELLED(0x010c) | QUIC     
+
+[RESET_STREAM](https://www.rfc-editor.org/rfc/rfc9000.html#name-reset_stream-frames) abruptly
+terminates sending on a stream. An important difference between HTTP/2 and
+HTTP/3 is frames are received out-of-order. Because of this, a `RESET_STREAM` sent after a
+completed response could still result in the response being aborted.
+
+HTTP/3 also has the QUIC `STOP_SENDING` frame. This frame is sent by the server when a
+stream's response side completes before the request side. Using `STOP_SENDING` alone isn't
+appropriate for a deadline exceeded, because both stream directions should be aborted.
+However, `STOP_SENDING` should be sent along with `RESET_STREAM` if the deadline is exceeded
+while the request side is in-progress.
+
+## Rationale
+
+HTTP/3 is an upcoming Internet Protocol. There needs to be a standardized agreement for how
+gRPC over HTTP/3 works to maintain interoperability in the gRPC eco-system.
+
+## Implementation
+
+A trial implementation in grpc-dotnet is underway. .NET 6 is adding preview support for
+HTTP/3 to the .NET server and client. grpc-dotnet leverages that underlying HTTP/3 support.
```

Comment:
@lidizheng what about doing some python asyncio prototyping!?  https://github.com/aiortc/aioquic

Reply:
I understand that building a MVP gRPC Python HTTP/3 on top of the `aioquic` library probably isn't that hard. Currently, gRPC Python uses gRPC Core as the transport layer, it's not that trivial to move existing code onto another transport library. I hope if we want to invest resources, we have a working architecture and aiming for long-term support.